### PR TITLE
[2.x] PHP 8.4 support

### PR DESCRIPTION
**Pull request by Jubeki**
Add PHP 8.4 to test matrix.

### CODE REVIEW

**Comment by driesvints** [comment]
I'd honestly like to wait until it does support php 8.4 so we don't have to use  platform-req=php

**Comment by Jubeki** [comment]
Yeah that's why I left it in Draft for now. Just wanted to see if something else's needed to be fixed.

**Comment by crynobone** [comment]
@Jubeki I made a draft PR #91 just to experiment around. I closed it once I'm done with testing and applies the changes here.

**Comment by crynobone** [comment]
CleanShot 2024-09-25 at 12 55 04


I'm curious what would cause this error on PHP 8.4 tests https://github.com/laravel/framework/actions/runs/11024490854/job/30617746097

**Comment by Jubeki** [comment]
@crynobone maybe this is a little helpful (I don't know much about serialising closures yet)

Following Code snippet causes the error in the test `ImplicitRouteBindingTest::test_it_can_resolve_the_implicit_backed_enum_route_bindings_for_the_given_route` (first one in your image):
https://github.com/laravel/framework/blob/b17b2bd8ebcb850a909fd046c380d075fd2f118b/src/Illuminate/Routing/RouteSignatureParameters.php#L19-L23

This is the value of `$action['uses']` for 

PHP 8.3:
```
O:55:"Laravel\SerializableClosure\UnsignedSerializableClosure":1:{s:12:"serializable";O:46:"Laravel\SerializableClosure\Serializers\Native":5:{s:3:"use";a:0:{}s:8:"function";s:114:"function (\Illuminate\Tests\Routing\CategoryBackedEnum $category) {\n
            return $category->value;\n
        }";s:5:"scope";s:49:"Illuminate\Tests\Routing\ImplicitRouteBindingTest";s:4:"this";N;s:4:"self";s:32:"00000000000001060000000000000000";}}
```

PHP 8.4:
```
O:55:"Laravel\SerializableClosure\UnsignedSerializableClosure":1:{s:12:"serializable";O:46:"Laravel\SerializableClosure\Serializers\Native":5:{s:3:"use";a:0:{}s:8:"function";s:123:"function (\{closure:Illuminate\Tests\Routing\CategoryBackedEnum $category) {\n
            return $category->value;\n
        }";s:5:"scope";s:49:"Illuminate\Tests\Routing\ImplicitRouteBindingTest";s:4:"this";N;s:4:"self";s:32:"00000000000001060000000000000000";}}
```

See the small difference `s:114` -> `s:123?` and `function (\Illuminate\Tests\Routing\CategoryBackedEnum $category)` -> `function (\{closure:Illuminate\Tests\Routing\CategoryBackedEnum $category)`

Not sure if this helpful, but maybe a starting point for further debugging.

**Comment by Jubeki** [comment]
@crynobone I think this PR is related:
- https://github.com/php/php-src/pull/13550

How to fix this:

https://github.com/laravel/serializable-closure/blob/43244bfd7161fdbce7419c76f9a776b7d705b333/src/Serializers/Native.php#L140-L142

Instead use the following:

```php
$code = $reflector->getCode();
$code = str_replace('\{closure:', '', $code);
```

In the end, I think this is a bug with PHP and should probably be reported.

File: src/Serializers/Native.php [hunk 1]
```diff
@@ -140,6 +140,11 @@ public function __serialize()
 
         $code = $reflector->getCode();
 
+
+        if (PHP_VERSION >= 8.4) {
+            $code = str_replace('\{closure:', '', $code);
+        }
+
```
Comment (Jubeki) [inline src/Serializers/Native.php:147]: This changes fixes the problem...

See Runs in https://github.com/laravel/framework/pull/52919

**Comment by takaram** [comment]
I submitted an issue to php-src and they changed the return value of `ReflectionFunction::getNamespaceName()`.
https://github.com/php/php-src/issues/16122

Here's a sample code of how the reflection methods related to closure namespace work in PHP8.4.
https://3v4l.org/srIno/rfc

Now `ReflectionFunction::getNamespaceName()` returns `""` for anonymous functions.
`$reflection->getClosureScopeClass()->getNamespaceName()` will still work only if the closure is created in a method. Otherwise the namespace seems unable to be retrieved in a simple way.

File: .github/workflows/tests.yml [hunk 2]
```diff
@@ -43,7 +46,12 @@ jobs:
           echo "::add-matcher::${{ runner.tool_cache }}/phpunit.json"
 
       - name: Install PHP dependencies
+        if: matrix.php != 8.4
         run: composer update --prefer-stable --no-interaction --no-progress --with="illuminate/support:^${{ matrix.laravel }}"
 
+      - name: Install PHP dependencies (== PHP 8.4)
```
Comment (crynobone) [inline .github/workflows/tests.yml:52]:
```suggestion
      - name: Install PHP dependencies (--ignore-platform-req)
```

**Comment by Jubeki** [comment]
@nunomaduro do you have an idea, why for example the enum/class `P\Tests\SerializerGlobalEnum` does not exist? Could this have something to do with the changes to PHP 8.4 regarding Reflection? (Not sure what exactly was changed)

Pest correctly detects the defined classes / enums in the other PHP versions, but somehow not in PHP 8.4.

Screenshot 2024-10-24 at 14 52 18

**Comment by crynobone** [comment]
@Jubeki Can you rebase this to `develop` branch.

**Comment by Jubeki** [comment]
@crynobone rebase was somehow annoying, I merged develop into this branch.

Alls Tests passing 😄

**Review by crynobone — APPROVED** [review]

**Pull request merged**